### PR TITLE
StorageAPI

### PR DESCRIPTION
**Pull request by ankhachatryan**
Added implementation and tests. Model stores local representations of Directory and File, in util Data is kind of the Database for the project and api is the implementation part. Client interaction will be done via SystemApi class

### CODE REVIEW

File: src/main/java/org/example/api/IFileApi.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package org.example.api;
+
+import org.example.model.Directory;
+import org.example.model.MyFile;
+
+public interface IFileApi {
```
Comment (achogrid) [inline src/main/java/org/example/api/IFileApi.java]: this is just a suggestions, there are different class name conventions and its up to the concrete team which to use, but I personally prefer this convention for interface class pairs: FileService(Interface), FileServiceImpl(Implementation class). When you read this it's like 100% clear that these to files are related.

File: src/main/java/org/example/api/SystemApi.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
+package org.example.api;
+import org.example.model.Directory;
+import org.example.model.MyFile;
+
+public class SystemApi {
```
Comment (achogrid) [inline src/main/java/org/example/api/SystemApi.java:10]: First of all I want you to research a bit about layered architecture. You will mostly see examples with web and database layers included but just pay attention to how conceptually those layers are distributed. 

Good blog to read on this:
https://medium.com/java-vault/layered-architecture-b2f4ebe8d587

After you read this blog, I want you to rethink the architecture of the system. It's really good you have service and Api classes separate, as you can see by layers we simply mean classes here. Currently you have 2 layers, SystemApi class represents presentation layer, and FileService class application(business) layer. There is 1 think which can improve readability for layers, besides class separation you could also put different layer classes in a separate packages. For example "api" package is good for api related classes, and you could just create "service" package and put service classes there.

File: src/main/java/org/example/api/SystemApi.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    private final IFileApi api = new FileService();
+
+    public MyFile addFile(Directory directory, MyFile myFile) {
```
Comment (achogrid) [inline src/main/java/org/example/api/SystemApi.java]: think about this class once again, why we need this presentation layer at all? Here if we look at this implementation, is it needed at all? you simply invoke service methods form here. There is no additional "value" or logic done in this class except calling service methods. In general even for simple method creation, it's just bad practice to create function that only calls some other function. If there is no additional value to it, why not just ignore this Api class and directly call service?

Now let talk why we actually need Presentation layer and therefore this Api class.
MyFile and Directory classes are core business logic classes. If you already checked above comments and read about layers, 1 of the main things about layers is that we keep isolation between layers. We don't want to expose core business logic classes which belong to service into presentation/api layer. For example imagine same situation with User creation. And you would have a User class which has username and password fields. This way everytime we created a new User we would return a User object as response containing password field as well, which is kinda sensitive data and we don't want that. This presentation lauyer is basically to receive some input "translate" this input to understandable "language" for service layer and when service return response to "transalte" it to api caller so that it contains only the data which is needed.

File: src/main/java/org/example/api/SystemApi.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        return api.addFile(directory, myFile);
+    }
+
+    public MyFile removeFile(Directory directory, String fileName) {
```
Comment (achogrid) [inline src/main/java/org/example/api/SystemApi.java]: we can rewrite parameter and response types of all these functions to be simpler. instead of directory it can be just a String path, instead of MyFile response it can be a new MyResponse object which will contain fileId(I'll explain this id concept later, in the comments), fileContent this could be byte[] or just String. and String enum errorCode. This error codes are needed to udentify exception situations, what if file was not found? what if some random exception happened in the code? what if file already exists with same name in the folder? we want user who calls this API know exactly what happened?  Therefore based on service response we construct Api response object which goes to the caller.

File: src/main/java/org/example/api/SystemApi.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+        return api.removeFile(directory, fileName);
+    }
+
+    public MyFile removeFile(String fileName) {
+        return api.removeFile(fileName);
+    }
+
+    public MyFile getFile(String fileName) {
```
Comment (achogrid) [inline src/main/java/org/example/api/SystemApi.java]: I think we need to add id/fileId field to the File. what if I have 2 files with same names in differet directories? This method seems like it will return any of them which it finds first. Filename can't be a unique identifier across all directories as 2 files with same names can exist in different directories. We can add id field which will be unique for each created file, you can have a look at UUID class and generate unique ids with it.

File: src/main/java/org/example/api/SystemApi.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        return api.getFile(fileName);
+    }
+
+    public String readFile(String fileName) {
+        return api.readFile(fileName);
+    }
+
+    public MyFile copyFile(Directory from, Directory target, String fileName) {
```
Comment (achogrid) [inline src/main/java/org/example/api/SystemApi.java]: updated parameters would look like this: String fromPath, String targetPath, String fileId

File: src/main/java/org/example/api/FileService.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
+package org.example.api;
+
+import org.example.model.Directory;
+import org.example.model.MyFile;
+import org.example.util.Data;
+
+public class FileService implements IFileApi {
+    public FileService() {}
```
Comment (achogrid) [inline src/main/java/org/example/api/FileService.java]: Do we need this? just curious doen't it work without empty constructor?

File: src/main/java/org/example/api/FileService.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+    @Override
+    public MyFile addFile(Directory directory, MyFile myFile) {
+        if(!Data.getSystem().contains(directory)) {
+            System.out.println("Directory doesn't exist");
+            return null;
```
Comment (achogrid) [inline src/main/java/org/example/api/FileService.java]: First of all let's finally get rid of println statements and switch to Exceptions, It would be great to create our custom Exception class that will be thrown in such cases with some errorCode and senseful error message. Why not use simple existing Exception class??? because our custom exception class gives us more flexibility we can add anything we need inside it. In general every application we work on has its own single or multiple exception classes which include many different fields.

File: src/test/java/org/example/api/FileServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
+package org.example.api;
+
+import org.example.model.Directory;
+import org.example.model.MyFile;
+import org.example.util.Data;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
+
+class FileServiceTest {
```
Comment (achogrid) [inline src/test/java/org/example/api/FileServiceTest.java:14]: As for testing now we have 2 possibillites, we can test Service layer and we can also test presentation layer. In general we test both. I want you test both. Service layer test separately and presentation layer tests separately. Let me point out some test cases:
1. create 2 files with same name in 1 directory.
2. create file in a directory which does not exist.
3. remove file from directory that does not exist.
4. remove file which does not exist.
5. copy file to a wrong location.
6. move file to a wrong location.
These are just couple of edge cases, in general you should be covering everything whole code, all lines and functionality you have with tests. This is called test coverage. You should run tests in intelij with "run with coverage" button. It will show you the analytics which lines of code are tested.
image

This read lines mean those lines are not tested with the tests, and percentages mean how much of your code is tested. In general we try to test as much as possible in real world. But in our case 100% coverage is achiavable, we don't have much complex things which can not be tested. I want you to reach 100% ocverage of you code with tests.
image

**Review by achogrid — CHANGES_REQUESTED** [review]

File: src/main/java/org/example/Exception/InvalidDirectoryException.java
Comment (achogrid) [inline src/main/java/org/example/Exception/InvalidDirectoryException.java:1]: Let's look at all these custom exception classes. They are semantically same, they only differ with error message. Maybe it's better to have only 1 MyFileException class which will contain error message, error code as enums for each concrete exception situation.

**Review by achogrid — CHANGES_REQUESTED** [review]

File: src/main/java/org/example/Exception/InvalidDirectoryException.java
Comment (ankhachatryan) [reply src/main/java/org/example/Exception/InvalidDirectoryException.java:1]: I've made necessary changes please check it

File: src/main/java/org/example/api/IFileApi.java (hunk 1 above)
Comment (ankhachatryan) [reply src/main/java/org/example/api/IFileApi.java]: Changed the naming

File: src/main/java/org/example/api/SystemApi.java (hunk 2 above)
Comment (ankhachatryan) [reply src/main/java/org/example/api/SystemApi.java:10]: Yeah surely I understood that part, but without having the proper knowledge, but on my own. I created new package for the api, also removed some methods (findDirectory and findFileByID) from Data class, and moved it to FileServiceImpl as obviously I shouldn't have implemented anything in Data layer

File: src/main/java/org/example/Service/FileServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+import org.example.util.Data;
+public class FileServiceImpl implements FileService {
+    @Override
+    public MyFile addFile(Directory directory, MyFile myFile) throws MyFileException {
+        if(!Data.getSystem().contains(directory)) {
+            throw new MyFileException(ExceptionMessage.DirectoryNotFound);
+        } else if(myFile == null) {
+            throw new MyFileException(ExceptionMessage.InvalidFile);
+        } else
+            for(MyFile file: directory.getFiles()) {
```
Comment (achogrid) [inline src/main/java/org/example/Service/FileServiceImpl.java:19]: Please use {} :) , it's much easier to read with braces than without

File: src/main/java/org/example/api/SystemApi.java (hunk 3 above)
Comment (icarus-77) [reply src/main/java/org/example/api/SystemApi.java]: I made some modifications to keep the logic, I didn't change the return types, because if I return Strings or byte arrays from my methods in that case readFile method becomes useless, but we obviously need that, so for users there is one more step to get the content by using readFile method

File: src/main/java/org/example/api/SystemApi.java (hunk 6 above)
Comment (ankhachatryan) [reply src/main/java/org/example/api/SystemApi.java]: Instead of String field, I changed it to FileID which is of type Long and is unique

File: src/main/java/org/example/api/SystemApi.java (hunk 5 above)
Comment (ankhachatryan) [reply src/main/java/org/example/api/SystemApi.java]: I didn't implement UUID as it was difficult for later usage for accessing from tests, I did the ID part with long values.

File: src/main/java/org/example/api/SystemApi.java (hunk 4 above)
Comment (ankhachatryan) [reply src/main/java/org/example/api/SystemApi.java]: I made some modifications to keep the logic, I didn't change the return types, because if I return Strings or byte arrays from my methods in that case readFile method becomes useless, but we obviously need that, so for users there is one more step to get the content by using readFile method

File: src/main/java/org/example/api/SystemApi.java (hunk 3 above)
Comment (ankhachatryan) [reply src/main/java/org/example/api/SystemApi.java]: Yeah sure I understand that part intuitively, and also as program was quite simple I didn't think about adding the necessary value to this layer. But currently changed that for easier user interaction

File: src/main/java/org/example/api/FileService.java (hunk 7 above)
Comment (ankhachatryan) [reply src/main/java/org/example/api/FileService.java]: It was automatically added and I didn't pay attention))
Removed already

File: src/main/java/org/example/api/FileService.java (hunk 8 above)
Comment (ankhachatryan) [reply src/main/java/org/example/api/FileService.java]: Yeah sure, implemented exceptions too

File: src/test/java/org/example/api/FileServiceTest.java (hunk 9 above)
Comment (ankhachatryan) [reply src/test/java/org/example/api/FileServiceTest.java:14]: Added edge cases, also you'll notice some Data.clearSystem() by the end of every test, it is because our Data is static and once it is created it keeps going through all methods and obviously for every method we don't need the modifications of the system from previous tests, that's why I solved that part with this method, which is not in use for the application, but for tests. Also coverage is now 100%, but I didn't implement tests for service yet. Honestly do we actually need it here, as I think if there would be some problem in service, the api will give the errors from there, testing both for me in this case is like testing same thing twice. Obviously if you want me to implement test for Service too, I'll do it, but I don't understand the reason for this case.

Comment (achogrid) [reply src/test/java/org/example/api/FileServiceTest.java:14]: Looks good to me as it is now, we will take care of that double testing later in other projects. You are right testing 1 of the components in our case is enough. As for clearSystem() method, generally I don't like idea to have code which is only used in tests, but again in our case it's enough, we will learn more as we proceed.

File: src/main/java/org/example/api/SystemApi.java (hunk 2 above)
Comment (achogrid) [reply src/main/java/org/example/api/SystemApi.java:10]: yep, we could add data layer as well, but we don't really have much with data here, so it's good as it is now.

File: src/main/java/org/example/api/SystemApi.java (hunk 5 above)
Comment (achogrid) [reply src/main/java/org/example/api/SystemApi.java]: I don't get why it was difficult to use UUID can you, tell a bit more about that?

File: src/main/java/org/example/model/Directory.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        MyFile temp = null;
+        for(MyFile myFile: files) {
+            if(myFile.getFileName().equals(fileName)) {
+                temp = myFile;
+                files.remove(myFile);
+                myFile.setDirectory(null);
+                break;
+            }
+        }
+        if(temp == null) {
```
Comment (achogrid) [inline src/main/java/org/example/model/Directory.java:50]: cleanup

**Review by achogrid — APPROVED** [review]
LGTM,
